### PR TITLE
Implement StoreVectorNxM for Arm64

### PR DESCRIPTION
**Pull request by SwapnilGaikwad**
Contribute towards #84510 

```c#
// ST3 (multiple structures)
public static unsafe void StoreVector128x3(byte*   address, (Vector128<byte>   Value1, Vector128<byte>   Value2, Vector128<byte>   Value3) value);
public static unsafe void StoreVector128x3(sbyte*  address, (Vector128<sbyte>  Value1, Vector128<sbyte>  Value2, Vector128<sbyte>  Value3) value);
public static unsafe void StoreVector128x3(short*  address, (Vector128<short>  Value1, Vector128<short>  Value2, Vector128<short>  Value3) value);
public static unsafe void StoreVector128x3(ushort* address, (Vector128<ushort> Value1, Vector128<ushort> Value2, Vector128<ushort> Value3) value);
public static unsafe void StoreVector128x3(int*    address, (Vector128<int>    Value1, Vector128<int>    Value2, Vector128<int>    Value3) value);
public static unsafe void StoreVector128x3(uint*   address, (Vector128<uint>   Value1, Vector128<uint>   Value2, Vector128<uint>   Value3) value);
public static unsafe void StoreVector128x3(long*   address, (Vector128<long>   Value1, Vector128<long>   Value2, Vector128<long>   Value3) value);
public static unsafe void StoreVector128x3(ulong*  address, (Vector128<ulong>  Value1, Vector128<ulong>  Value2, Vector128<ulong>  Value3) value);
public static unsafe void StoreVector128x3(float*  address, (Vector128<float>  Value1, Vector128<float>  Value2, Vector128<float>  Value3) value);
public static unsafe void StoreVector128x3(double* address, (Vector128<double> Value1, Vector128<double> Value2, Vector128<double> Value3) value);

public static unsafe void StoreVector64x3(byte*   address, (Vector64<byte>   Value1, Vector64<byte>   Value2, Vector64<byte>   Value3) value);
public static unsafe void StoreVector64x3(sbyte*  address, (Vector64<sbyte>  Value1, Vector64<sbyte>  Value2, Vector64<sbyte>  Value3) value);
public static unsafe void StoreVector64x3(short*  address, (Vector64<short>  Value1, Vector64<short>  Value2, Vector64<short>  Value3) value);
public static unsafe void StoreVector64x3(ushort* address, (Vector64<ushort> Value1, Vector64<ushort> Value2, Vector64<ushort> Value3) value);
public static unsafe void StoreVector64x3(int*    address, (Vector64<int>    Value1, Vector64<int>    Value2, Vector64<int>    Value3) value);
public static unsafe void StoreVector64x3(uint*   address, (Vector64<uint>   Value1, Vector64<uint>   Value2, Vector64<uint>   Value3) value);
public static unsafe void StoreVector64x3(float*  address, (Vector64<float>  Value1, Vector64<float>  Value2, Vector64<float>  Value3) value);

// ST4 (multiple structures)
public static unsafe void StoreVector128x4(byte*   address, (Vector128<byte>   Value1, Vector128<byte>   Value2, Vector128<byte>   Value3, Vector128<byte>   Value4) value);
public static unsafe void StoreVector128x4(sbyte*  address, (Vector128<sbyte>  Value1, Vector128<sbyte>  Value2, Vector128<sbyte>  Value3, Vector128<sbyte>  Value4) value);
public static unsafe void StoreVector128x4(short*  address, (Vector128<short>  Value1, Vector128<short>  Value2, Vector128<short>  Value3, Vector128<short>  Value4) value);
public static unsafe void StoreVector128x4(ushort* address, (Vector128<ushort> Value1, Vector128<ushort> Value2, Vector128<ushort> Value3, Vector128<ushort> Value4) value);
public static unsafe void StoreVector128x4(int*    address, (Vector128<int>    Value1, Vector128<int>    Value2, Vector128<int>    Value3, Vector128<int>    Value4) value);
public static unsafe void StoreVector128x4(uint*   address, (Vector128<uint>   Value1, Vector128<uint>   Value2, Vector128<uint>   Value3, Vector128<uint>   Value4) value);
public static unsafe void StoreVector128x4(long*   address, (Vector128<long>   Value1, Vector128<long>   Value2, Vector128<long>   Value3, Vector128<long>   Value4) value);
public static unsafe void StoreVector128x4(ulong*  address, (Vector128<ulong>  Value1, Vector128<ulong>  Value2, Vector128<ulong>  Value3, Vector128<ulong>  Value4) value);
public static unsafe void StoreVector128x4(float*  address, (Vector128<float>  Value1, Vector128<float>  Value2, Vector128<float>  Value3, Vector128<float>  Value4) value);
public static unsafe void StoreVector128x4(double* address, (Vector128<double> Value1, Vector128<double> Value2, Vector128<double> Value3, Vector128<double> Value4) value);

public static unsafe void StoreVector64x4(byte*   address, (Vector64<byte>   Value1, Vector64<byte>   Value2, Vector64<byte>   Value3, Vector64<byte>   Value4) value);
public static unsafe void StoreVector64x4(sbyte*  address, (Vector64<sbyte>  Value1, Vector64<sbyte>  Value2, Vector64<sbyte>  Value3, Vector64<sbyte>  Value4) value);
public static unsafe void StoreVector64x4(short*  address, (Vector64<short>  Value1, Vector64<short>  Value2, Vector64<short>  Value3, Vector64<short>  Value4) value);
public static unsafe void StoreVector64x4(ushort* address, (Vector64<ushort> Value1, Vector64<ushort> Value2, Vector64<ushort> Value3, Vector64<ushort> Value4) value);
public static unsafe void StoreVector64x4(int*    address, (Vector64<int>    Value1, Vector64<int>    Value2, Vector64<int>    Value3, Vector64<int>    Value4) value);
public static unsafe void StoreVector64x4(uint*   address, (Vector64<uint>   Value1, Vector64<uint>   Value2, Vector64<uint>   Value3, Vector64<uint>   Value4) value);
public static unsafe void StoreVector64x4(float*  address, (Vector64<float>  Value1, Vector64<float>  Value2, Vector64<float>  Value3, Vector64<float>  Value4) value);

```

### CODE REVIEW

**Comment by SwapnilGaikwad** [comment]
Hi @kunalspathak , would you prefer an overloaded name for the interleaved multi-structure stores like we did for `StoreSelectedScalar`? If yes then what should we call such methods? `Store` or `StoreVector` may not convey the interleaving nature of underlying store instructions. 🤔

**Comment by kunalspathak** [comment]
> for the interleaved multi-structure stores 

sorry, but could you confirm which APIs are you asking about?

**Comment by SwapnilGaikwad** [comment]
> > for the interleaved multi-structure stores
> 
> sorry, but could you confirm which APIs are you asking about?

The StoreVectorNX2, StoreVectorNx3, StoreVectorNx3 that result in ST2, ST3 and ST4 instructions respectively.

**Comment by SwapnilGaikwad** [comment]
Could you please confirm if my understanding is correct?
- LoadVector(N)x(M) is mapped to LD(M). E.g., LoadVector128x3 to LD3
- StoreVector(N)x(M) should map to ST(M), E.g., StoreVector128x3 to ST3

- StoreVector(N)x(M)AndZip should map to ST1 with M registers.

If this is correct, then aren't these names a little confusing?
LD2/3/4 would unzip/de-interleave while loading the input registers while LD1 will load given input vectors 2/3/4 consecutively.
Thus, wouldn't it make more sense for LoadVector(N)x(M) to emit LD1 with M input values and vice-versa for LoadVector(N)x(M)AndZip?

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:820]: i think we should validate the number of registers too for each of x2, x3 and x4.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:1691]: TestName should be fixed to `x3` and `x4`.

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Changes looks good, some minor fixes needed.

**Comment by kunalspathak** [comment]
> Thus, wouldn't it make more sense for LoadVector(N)x(M) to emit LD1 with M input values and vice-versa for LoadVector(N)x(M)AndZip?

Looking carefully, yes, you are right. I swapped the two. I will send a PR to fix it. Thanks for spotting it.

**Comment by SwapnilGaikwad** [comment]
> > Thus, wouldn't it make more sense for LoadVector(N)x(M) to emit LD1 with M input values and vice-versa for LoadVector(N)x(M)AndZip?
> 
> Looking carefully, yes, you are right. I swapped the two. I will send a PR to fix it. Thanks for spotting it.

Cool, I'll update this PR accordingly 👍

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp
Comment (SwapnilGaikwad) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:820]: Done

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs
Comment (SwapnilGaikwad) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs]: Done

**Comment by SwapnilGaikwad** [comment]
After merging #93223, I'll rebase/merge this PR and then mark it ready for review.

**Comment by SwapnilGaikwad** [comment]
> After merging #93223, I'll rebase/merge this PR and then mark it ready for review.

As the #93223 taking longer, I'll mark this PR ready for review. If this can progress further quickly then we can merge this. I'll rebase/merge these changes wherever necessary.

**Review by kunalspathak — COMMENTED** [review]
LGTM. Just a request to update the documentation for LoadVector equivalent.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs
Comment (SwapnilGaikwad) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs:3778]: Sure, I'll update this shortly.

Comment (SwapnilGaikwad) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs:3778]: Done!

Comment (kunalspathak) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs:3778]: Thank you. I will merge this once CI is green.

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**
